### PR TITLE
[LTC] Enable CI for building and testing

### LINKED ISSUE #63610: Enable some amount of CI tests on lazy_tensor_staging branch
Tests to consider adding are:
1) test_ptltc, C++ unit tests.
2) example.py (https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/lazy_tensor_core/example.py). Needed to cover some python scripts.
3) test_ops.py (https://github.com/pytorch/xla/blob/master/test/test_ops.py). Needs to change the device to lazy and run it with the TS backend.
4) other tests in the above folder could be useful too, like training MNIST end to end. This one should be super fast even for CPU.

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This patch enables PyTorch CI for builiding and testing LazyTensorCore.

The initial attempt is to build LTC for linux-xenial-cuda11.1 environments and run test_ptltc C++ unit tests in the test1 bots.

Test Plan:
PyTorch CI \*linux-xenial-cuda11.1\*-test1\*.

Fixes #63610.

### CODE REVIEW

File: .jenkins/pytorch/build.sh [hunk 1]
```diff
@@ -322,3 +322,18 @@ if [[ "$BUILD_ENVIRONMENT" != *libtorch* && "$BUILD_ENVIRONMENT" != *bazel* ]];
   # don't do this for libtorch as libtorch is C++ only and thus won't have python tests run on its build
   python test/run_test.py --export-past-test-times
 fi
+
+# Test Lazy Tensor Core
+# No particular reasons why this environment, just restrict lazy tensor to one to limit potential breakages.
+if [[ "$BUILD_ENVIRONMENT" == *linux-xenial-cuda11.1* ]]; then
+  pushd lazy_tensor_core/
+  ./scripts/apply_patches.sh
```
Comment (malfet) [inline .jenkins/pytorch/build.sh:330]: Having code in the same repo exist as set of patches feels a bit weird
Plus, it puts unreasonable burden on all developers to update those patches when they modify the code
Perhaps `#ifdefs` in the code would be better

**Review by malfet — CHANGES_REQUESTED** [review]
I don't have enough context, but it looks like you are trying to hijack linux-xenial-cuda11.1 build and test to run LazyTensor tests, which is:
 - Fragile, test will get disabled without notifying anyone
 - Confusing to devs, who think the only difference between say 10.2 and 11.1 config is cuda version
 - Reduces test coverage/weakens signal, since same config are used to test the same thing
If you need to test something custom, please add new build/test configs to the file
Also, when adding a new config it would be great(to avoid ParallelTBB fiasco) to share a bit longer RFC that would explain to everybody what this config is doing.

**Comment by wconstab** [comment]
@malfet for some context:

We have a branch `lazy_tensor_staging` which is long-lived but eventually we'll kill it and merge things to master.  For now, we want to enable CI _on that branch only_ for our own sanity before we manage to merge to master.  Once things are on master, they will look very different, including
* the build/test flow will be more native and included in pytorch's usual cmake and unittest suite
* the apply_patches thing will be gone

For now, i don't care too much if we run the tests on one particular config or on many.  But i want @alanwaketan to not have to spend too much time setting up the short-lived infra.  

Do you have a suggestion that is lightweight, just to enable testing on our branch?

File: .jenkins/pytorch/build.sh (hunk 1 above)
Comment (alanwaketan) [reply .jenkins/pytorch/build.sh:330]: Those are all good reasons. However, Will has filled up some contexts here. Wondering if you have any suggestions how to make our use case viable.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp [hunk 2]
```diff
@@ -3,7 +3,7 @@
 #include <ATen/core/Reduction.h>
 #include <torch/csrc/jit/frontend/sugared_value.h>
 #include <torch/jit.h>
-#include <torch/torch.h>
+// #include <torch/torch.h>
```
Comment (alanwaketan) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:6]: Yesterday I saw a weird bug in the CI that <torch/torch.h> and <torch/csrc/jit/frontend/sugared_value.h> point to the same file from different install locations, and therefore introduce a redefinition bug. That's why I comment out this header to see if it fixes the issue. Here is a chopped version:
```
Sep 15 01:10:26 In file included from /opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include/torch/nn/module.h:6:0,
Sep 15 01:10:26                  from /opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include/torch/nn/cloneable.h:3,
Sep 15 01:10:26                  from /opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include/torch/nn.h:3,
Sep 15 01:10:26                  from /opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include/torch/all.h:13,
Sep 15 01:10:26                  from /opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include/torch/torch.h:3,
Sep 15 01:10:26                  from lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:6:
Sep 15 01:10:26 /opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include/torch/ordered_dict.h:13:7: error: redefinition of 'class torch::OrderedDict<Key, Value>'
Sep 15 01:10:26  class OrderedDict {
Sep 15 01:10:26        ^~~~~~~~~~~
Sep 15 01:10:26 In file included from /var/lib/jenkins/workspace/torch/csrc/jit/api/module.h:13:0,
Sep 15 01:10:26                  from /var/lib/jenkins/workspace/torch/csrc/jit/frontend/sugared_value.h:8,
Sep 15 01:10:26                  from lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:4:
Sep 15 01:10:26 /var/lib/jenkins/workspace/torch/csrc/api/include/torch/ordered_dict.h:13:7: note: previous definition of 'class torch::OrderedDict<Key, Value>'
Sep 15 01:10:26  class OrderedDict {
``` 

It's weird that <torch/torch.h> is installed at `/opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include/torch/torch.h` while <torch/csrc/jit/frontend/sugared_value.h> is installed at `/var/lib/jenkins/workspace/torch/csrc/jit/frontend/sugared_value.h`. Wondering if someone can explain this situation to me and decide if the deleting <torch/torch.h> is a good solution.

P.S. A detailed report: https://app.circleci.com/pipelines/github/pytorch/pytorch/380340/workflows/84296af5-fee4-4812-95d3-419e097047a6/jobs/16023443.

**Review by malfet — APPROVED** [review]
Haven't realized that PR targets `lazy_tensor_staging`. In that case, it looks good to me, you guys are free to do whatever you want

File: .jenkins/pytorch/build.sh (hunk 1 above)
Comment (wconstab) [reply .jenkins/pytorch/build.sh:330]: This is OK for now. we can do it better once we refactor the lazy tensor build system for merging lazy tensor to master.

**Review by wconstab — APPROVED** [review]

**Comment by alanwaketan** [comment]
Thanks @wconstab and @malfet for approving this patch. The other CUDA environment, i.e., linux_bionic_cuda10_2_cudnn7_py3_9_gcc7, uses Python3.9 which doesn't work with the current lazy tensor building process. Thus, let's restrict it to one environment.

**Comment by alanwaketan** [comment]
Land the commit as it is since the failures don't seem to be related.

**Pull request merged**